File: range_sensor/range_sensor/counter_with_delay_action_server.py
#! /usr/bin/env python

# This code has been adapted from the ROS Wiki actionlib tutorials to the context
# of this course.
# (http://wiki.ros.org/range_sensor_msgs/Tutorials)

import time

import rclpy
from rclpy.action import ActionServer
from rclpy.node import Node

from range_sensors_interfaces.action import CounterWithDelay

class CounterWithDelayServer(Node):

    def __init__(self):
        super().__init__('counter_with_delay_action_server')
        self._action_name = 'counter_with_delay_action_server'
        self._action_server = ActionServer(
            self,
            CounterWithDelay,
            self._action_name,
            self.action_server_callback)

    def action_server_callback(self, goal_handle):
        self.get_logger().info('Executing goal...')

        ## Assignment 3 - Part3
        ## modify counter delay using a private parameter .
        ## Uncomment the following lines (37-41) and modify them acordingly

        # Declare the parameter with a default value if it doesn't exist
        self.declare_parameter('counter_delay', 1.0)
        
        # Get the parameter value
        counter_delay_value = self.get_parameter('counter_delay').value
        
        # Check if the parameter was set explicitly or is using the default
        if self.get_parameter('counter_delay').get_parameter_value().type == 4:  # 4 means PARAMETER_NOT_SET
            self.get_logger().info(
                "Parameter 'counter_delay' not found on the parameter server. Using default value of 1.0s for counter delay."
            )
            counter_delay_value = 1.0
        else:
            self.get_logger().info(
                f"Parameter 'counter_delay' was found on the parameter server. Using {counter_delay_value}s for counter delay."
            )


        ##  End of Assignment 3 - Part3

        # Variable for delay
        #delay_rate = rospy.Rate(counter_delay_value)
        delay_rate = 100

        # Variable to decide the final state of the action server.
        #success = True

        # publish info to the console for the user
        self.get_logger().info('%s action server is counting up to  %i with %fs delay between each count' % (self._action_name, goal_handle.request.num_counts, counter_delay_value))

        feedback = CounterWithDelay.Feedback()

        # start executing the action
        for counter_idx in range(0, goal_handle.request.num_counts):
            # check that preempt has not been requested by the client

            # publish the feedback
            feedback.counts_elapsed = counter_idx
            goal_handle.publish_feedback(feedback)
            # Wait for counter_delay s before incrementing the counter.
            time.sleep(1)

        goal_handle.succeed()

        self.get_logger().info('%s: Succeeded' % self._action_name)
        result = CounterWithDelay.Result()
        result.result_message = "Successfully completed counting."
        return result


def main(args=None):
    rclpy.init(args=args)

    counter_with_delay_server = CounterWithDelayServer()

    rclpy.spin(counter_with_delay_server)


if __name__ == '__main__':
    main()

File: range_sensor_cpp/src/counter_with_delay_action_server.cpp
#include <memory>
#include <chrono>
#include <string>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "range_sensors_interfaces/action/counter_with_delay.hpp"

using namespace std::chrono_literals;

class CounterWithDelayServer : public rclcpp::Node {
public:
    using CounterWithDelay = range_sensors_interfaces::action::CounterWithDelay;
    using GoalHandleCounterWithDelay = rclcpp_action::ServerGoalHandle<CounterWithDelay>;

  explicit CounterWithDelayServer(const rclcpp::NodeOptions & options = rclcpp::NodeOptions())
  : Node("counter_with_delay_action_server", options)
  {
    using namespace std::placeholders;
#if 0
    this->action_server_ = rclcpp_action::create_server<CounterWithDelay>(
      this,
      "counter_with_delay_action_server",
      std::bind(&CounterWithDelayServer::handle_goal, this, std::placeholders::_1, std::placeholders::_2),
      std::bind(&CounterWithDelayServer::handle_cancel, this, std::placeholders::_1),
      std::bind(&CounterWithDelayServer::handle_accepted, this, std::placeholders::_1));
#endif
  }

  
private:
    rclcpp_action::Server<CounterWithDelay>::SharedPtr action_server_;

    rclcpp_action::GoalResponse handle_goal(const std::shared_ptr<GoalHandleCounterWithDelay> goal_handle)
    {
        RCLCPP_INFO(this->get_logger(), "Received goal request with number of counts %d", goal_handle->get_goal()->num_counts);
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    void handle_accepted(const std::shared_ptr<GoalHandleCounterWithDelay> goal_handle) {
        std::thread{std::bind(&CounterWithDelayServer::execute, this, std::placeholders::_1), goal_handle}.detach();
    }

    void execute(const std::shared_ptr<GoalHandleCounterWithDelay> goal_handle){
        RCLCPP_INFO(this->get_logger(), "Executing goal...");

        // Get the parameter value
        double counter_delay_value;
        this->get_parameter("counter_delay", counter_delay_value);

        RCLCPP_INFO(this->get_logger(), "Using %f seconds for counter delay.", counter_delay_value);

        // Variable to store feedback
        auto feedback = std::make_shared<CounterWithDelay::Feedback>();
        auto result = std::make_shared<CounterWithDelay::Result>();

        // Start executing the action
        for (int counter_idx = 0; counter_idx < goal_handle->get_goal()->num_counts; ++counter_idx) {
            // Check if the goal has been canceled
            if (goal_handle->is_canceling()) {
                result->result_message = "Canceled counting.";
                goal_handle->canceled(result);
                RCLCPP_INFO(this->get_logger(), "Goal canceled");
                return;
            }

            // Publish feedback
            feedback->counts_elapsed = counter_idx;
            goal_handle->publish_feedback(feedback);

            // Wait for counter_delay seconds before incrementing the counter
            std::this_thread::sleep_for(std::chrono::duration<double>(counter_delay_value));
        }

        // Check if goal is done
        if (rclcpp::ok()) {
            RCLCPP_INFO(this->get_logger(), "Succeeded");
            result->result_message = "Successfully completed counting.";
            goal_handle->succeed(result);
        }
    }
    rclcpp_action::CancelResponse handle_cancel(const std::shared_ptr<GoalHandleCounterWithDelay> goal_handle){
        RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

};

int main(int argc, char * argv[]) {
    rclcpp::init(argc, argv);
    auto counter_with_delay_server = std::make_shared<CounterWithDelayServer>();
    rclcpp::spin(counter_with_delay_server);
    rclcpp::shutdown();
    return 0;
}

